### PR TITLE
Restore partial support of RuleQuery in RuleFinder to fix issue with Gro...

### PR DESCRIPTION
**Pull request by henryju**
...ovy plugin

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/rules/RuleFinder.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
 
 /**
  * @since 2.3
- * @deprecated since 5.1 DB access will soon be removed from batch side. Use {@link ActiveRules} instead.
+ * @deprecated since 5.1. Use {@link ActiveRules} on batch side.
  */
 @Deprecated
 public interface RuleFinder extends TaskComponent, ServerComponent {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/rules/RuleFinder.java]: INFODo not forget to remove this deprecated code someday.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1133)

File: sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java [hunk 2]
```diff
@@ -48,18 +63,48 @@ public Rule findByKey(String repositoryKey, String key) {
...
+      throw new IllegalArgumentException("Non unique result for rule query: " + ReflectionToStringBuilder.toString(query, ToStringStyle.SHORT_PREFIX_STYLE));
+    } else if (all.size() == 0) {
+      return null;
+    } else {
+      return all.iterator().next();
+    }
   }
 
   @Override
   public Collection<Rule> findAll(RuleQuery query) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java]: :new: MAJORThe Cyclomatic Complexity of this method "findAll" is 13 which is greater than 10 authorized.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java [hunk 3]
```diff
@@ -26,8 +31,18 @@
...
 
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 
+/**
+ * FIXME Waiting for the list of all server rules on batch side this is implemented by redirecting on ActiveRules. This is not correct
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java]: :new: MAJORTake the required action to fix the issue indicated by this comment.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)

File: sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java [hunk 4]
```diff
@@ -26,8 +31,18 @@
...
+ * since there is a difference between a rule that doesn't exists and a rule that is not activated in project quality profile.
+ *
+ */
 public class RuleFinderCompatibility implements RuleFinder {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java]: MAJOR"RuleFinder" is deprecated, implement the suggested replacement instead.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1724)

File: sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java [hunk 5]
```diff
@@ -48,18 +63,48 @@ public Rule findByKey(String repositoryKey, String key) {
...
+    return toRule(activeRules.find(key));
   }
 
   @Override
   public Rule find(RuleQuery query) {
-    throw new UnsupportedOperationException("Unable to find rule by query");
+    Collection<Rule> all = findAll(query);
+    if (all.size() > 1) {
+      throw new IllegalArgumentException("Non unique result for rule query: " + ReflectionToStringBuilder.toString(query, ToStringStyle.SHORT_PREFIX_STYLE));
+    } else if (all.size() == 0) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java]: :new: MAJORUse isEmpty() to check whether the collection is empty or not.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1155)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 new issues:
* 3 major

Details:
* [The Cyclomatic Complexity of this method "findAll" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/a8dd14a9695351eff7a1d0aab67de1bea93fbd04/sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java#L82)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* [Take the required action to fix the issue indicated by this comment.](https://github.com/SonarSource/sonarqube/blob/a8dd14a9695351eff7a1d0aab67de1bea93fbd04/sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java#L42)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)
* [Use isEmpty() to check whether the collection is empty or not.](https://github.com/SonarSource/sonarqube/blob/a8dd14a9695351eff7a1d0aab67de1bea93fbd04/sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java#L74)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1155)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

Details:
* [The Cyclomatic Complexity of this method "findAll" is 13 which is greater than 10 authorized.](https://github.com/SonarSource/sonarqube/blob/b4c8a3a588742ac3c231f71fcad8a735c39fcf9b/sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java#L82)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
* [Take the required action to fix the issue indicated by this comment.](https://github.com/SonarSource/sonarqube/blob/b4c8a3a588742ac3c231f71fcad8a735c39fcf9b/sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java#L42)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)

**Comment by simonbrandhof** [comment]
LGTM

**Pull request merged**